### ISSUE TITLE
[prefer-snapshot-hint] false positives when providing snapshot hint as first parameter

### ISSUE THREAD

**Issue by lostfictions** [jest-community/eslint-plugin-jest#1068]
Hey there, thanks for adding the `prefer-snapshot-hint` rule -- it's a great idea! Unfortunately, it doesn't seem to work when the hint is passed in as the first parameter.

Here's a practical motivating example:

```ts
test("typescript errors haven't increased", async () => {
  const { stdout } = await execa("tsc", { reject: false });
  const errors = stdout
    .split("\n")
    .filter((line) => line.startsWith("src/"))
    .map((line) => line.replace(/^(src\/[^(]+)(?:\([\d,]+\))/, "$1"));

  expect(errors.length).toMatchSnapshot("tsc error count");
  //                    ^^^^^^^^^^^^^^^
  //                    yields a `prefer-snapshot-hint` warning
  expect(errors).toMatchSnapshot("tsc errors");
  //             ^^^^^^^^^^^^^^^
  //             ditto
});
```

`errors` is an array, and `errors.length` is obviously a number. I can't do something like `expect(errors.length).toMatchSnapshot({}, "my hint text")` or `expect(errors.length).toMatchSnapshot(undefined, "my type hint")` -- I either have to provide a real matcher (which in the case of `errors.length` has to be a number) or I have to omit the matcher entirely, as I've done in the above snippet.

(Note that this is reflected in `@types/jest` as well -- it correctly infers the matcher type for `.toMatchSnapshot` and yields a typechecking error if I try to use `undefined` or `{}` as a matcher when I `expect()` a number, but it does allow a hint to be passed as the first parameter. [Here's the relevant lines](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/277df65eb7d3bad751f87fdc1dcfb9e7205bd46a/types/jest/index.d.ts#L952-L962) on DefinitelyTyped.) 

The `errors` array itself fares no better -- I can't use `{}` as a matcher, and `[]` isn't semantically the same as no matcher at all (it instead means "match an empty array").

The only correct way to add a snapshot hint for these calls is to pass it as the first parameter. The upshot is that `prefer-snapshot-hint` is effectively broken right now for any `expect` that doesn't wrap an object. 

Here's a minimal repro:

```js
/* eslint jest/prefer-snapshot-hint: ["error", "always"] */

describe("repro", () => {
  it("yields a false positive for `prefer-snapshot-hint`", () => {
    expect(1).toMatchSnapshot("a very descriptive snapshot hint")
  });
})
```

**Comment by G-Rath** [comment]
huh, ok TIL - it doesn't seem to be mentioned anywhere in the [docs](https://jestjs.io/docs/next/expect#tomatchsnapshotpropertymatchers-hint) that I can find.

(iirc this actually should make the code a bit simpler too)
